### PR TITLE
Fix SymInt Scalar equality for pow backward

### LINKED ISSUE #185715: INTERNAL ASSERT FAILED: NYI SymInt equality in c10/core/Scalar.h when using torch.compile(dynamic=True) and torch.autograd.grad with torch.pow
### 🐛 Describe the bug

When compiling a function using `torch.compile` with `dynamic=True` and taking gradients via `torch.autograd.grad` after a `torch.pow` operation, it throws an internal assert error in the C++ engine: `false INTERNAL ASSERT FAILED at "/__w/pytorch/pytorch/c10/core/Scalar.h":273, please report a bug to PyTorch. NYI SymInt equality`. 

This occurs on both `aot_eager` and `inductor` backends. And maybe the issue only reproduces when the exponent is passed as a Python int argument. If the exponent is written as a literal constant inside the function, e.g. torch.pow(x, 0), the issue does not reproduce.

### Reproduction Script
```python
import torch
import traceback

print("torch", torch.__version__)
torch._dynamo.config.suppress_errors = False

def fn(x, exponent):
    y = torch.pow(x, exponent)
    loss = y.sum()
    grad, = torch.autograd.grad(loss, x)
    return y, grad

x = torch.ones((1, 3), device="cuda", dtype=torch.bfloat16, requires_grad=True)
exponent = 0  

print("\n=== eager ===")
y, g = fn(x, exponent)
print("y:", y)
print("grad:", g)

for backend in ["aot_eager", "inductor"]:
    print(f"\n=== torch.compile backend={backend}, dynamic=True ===")
    compiled_fn = torch.compile(fn, backend=backend, dynamic=True)

    x = torch.ones((1, 3), device="cuda", dtype=torch.bfloat16, requires_grad=True)

    try:
        y, g = compiled_fn(x, exponent)
        print("y:", y)
        print("grad:", g)
    except Exception:
        traceback.print_exc()
```

### Error log

[errorlog.txt](https://github.com/user-attachments/files/28425435/errorlog.txt)

### Versions

PyTorch version: 2.13.0.dev20260521+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.0.140
Nvidia driver version: 596.49
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_tensor_ir.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.21.1
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A
ersions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260521+cu130
[pip3] torchaudio==2.11.0.dev20260525+cu130
[pip3] torchvision==0.28.0.dev20260525+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] numpy 2.2.6 pypi_0 pypi
[conda] nvidia-cublas 13.1.1.3 pypi_0 pypi
[conda] nvidia-cuda-cupti 13.0.85 pypi_0 pypi
[conda] nvidia-cuda-nvrtc 13.0.88 pypi_0 pypi
[conda] nvidia-cuda-runtime 13.0.96 pypi_0 pypi
[conda] nvidia-cudnn-cu13 9.20.0.48 pypi_0 pypi
[conda] nvidia-cufft 12.0.0.61 pypi_0 pypi
[conda] nvidia-curand 10.4.0.35 pypi_0 pypi
[conda] nvidia-cusolver 12.0.4.66 pypi_0 pypi
[conda] nvidia-cusparse 12.6.3.3 pypi_0 pypi
[conda] nvidia-cusparselt-cu13 0.8.1 pypi_0 pypi
[conda] nvidia-nccl-cu13 2.29.7 pypi_0 pypi
[conda] nvidia-nvjitlink 13.0.88 pypi_0 pypi
[conda] nvidia-nvtx 13.0.85 pypi_0 pypi
[conda] torch 2.13.0.dev20260521+cu130 pypi_0 pypi
[conda] torchaudio 2.11.0.dev20260525+cu130 pypi_0 pypi
[conda] torchvision 0.28.0.dev20260525+cu130 pypi_0 pypi
[conda] triton 3.7.0+git88b227e2 pypi_0 pypi

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @chauhang @penguinwu @aditvenk @laithsakka

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185849

When torch.compile(dynamic=True) traces a Python integer argument, AOTAutograd can save it as a symbolic integer Scalar. The scalar-exponent pow backward formula checks whether the exponent equals zero, but Scalar::equal asserted on HAS_si values with "NYI SymInt equality" instead of comparing them. That made torch.autograd.grad after torch.pow crash during compilation for dynamic Python int exponents.

Implement Scalar::equal for symbolic integer Scalars by converting comparable numeric constants to int64_t and comparing with SymInt::sym_eq(...).guard_bool(...). Non-integral, non-finite, or out-of-range numeric constants return false before any integer cast, avoiding undefined behavior around the int64_t limits. Complex comparisons use the same path only when the imaginary part is zero.

I considered special-casing pow backward, but the missing operation is Scalar equality for symbolic integers; fixing it in Scalar keeps the behavior consistent for other users of symbolic Scalar equality. The small uint64_t constructor annotation/update is lint-only cleanup required because touching this header brings existing clang-tidy checks into the changed-file lint set.

Fixes #185715
Generated by my agent

Test Plan:
- lintrunner -a
- ninja -C build torch_python -j 8
- python test/dynamo/test_aot_autograd.py -k test_autograd_grad_pow_python_int_exponent_dynamic
- lintrunner -a && ninja -C build c10_Scalar_test -j 8 && build/bin/c10_Scalar_test --gtest_filter=ScalarTest.SymIntEquality

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this draft PR is superseded by the maintainer-owned fix in https://github.com/pytorch/pytorch/pull/185851, which targets the same issue with the accepted autograd-level approach.
